From 52bf4c4bd2ae60bb90dc6296074e3830a23509d7 Mon Sep 17 00:00:00 2001
From: ditternation <ditternation@outlook.com>
Date: Sun, 21 Jun 2026 18:11:48 +0200
Subject: [PATCH] Introduce PlayIntegrityHooks

Preserve Android 17 task locking, tracing and cleanup. Only bypass the task
permission for the existing UN1CA hook's approved caller.

---
 .../wm/ActivityTaskManagerService.smali | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/smali_classes2/com/android/server/wm/ActivityTaskManagerService.smali b/smali_classes2/com/android/server/wm/ActivityTaskManagerService.smali
--- a/smali_classes2/com/android/server/wm/ActivityTaskManagerService.smali
+++ b/smali_classes2/com/android/server/wm/ActivityTaskManagerService.smali
@@ -8377,9 +8377,19 @@
 .method public final getFocusedRootTaskInfo()Landroid/app/ActivityTaskManager$RootTaskInfo;
     .locals 5
 
+    iget-object v0, p0, Lcom/android/server/wm/ActivityTaskManagerService;->mContext:Landroid/content/Context;
+
+    invoke-static {v0}, Lio/mesalabs/unica/PlayIntegrityHooks;->shouldBypassTaskPermission(Landroid/content/Context;)Z
+
+    move-result v0
+
+    if-nez v0, :unica_task_permission
+
     const-string/jumbo v0, "getFocusedRootTaskInfo()"
 
     invoke-static {v0}, Lcom/android/server/wm/ActivityTaskManagerService;->enforceTaskPermission(Ljava/lang/String;)V
 
+    :unica_task_permission
+
     invoke-static {}, Landroid/os/Binder;->clearCallingIdentity()J
 
@@ -12876,9 +12886,19 @@
 .method public final registerTaskStackListener(Landroid/app/ITaskStackListener;)V
     .locals 2
 
+    iget-object v0, p0, Lcom/android/server/wm/ActivityTaskManagerService;->mContext:Landroid/content/Context;
+
+    invoke-static {v0}, Lio/mesalabs/unica/PlayIntegrityHooks;->shouldBypassTaskPermission(Landroid/content/Context;)Z
+
+    move-result v0
+
+    if-nez v0, :unica_task_permission
+
     const-string/jumbo v0, "registerTaskStackListener()"
 
     invoke-static {v0}, Lcom/android/server/wm/ActivityTaskManagerService;->enforceTaskPermission(Ljava/lang/String;)V
 
+    :unica_task_permission
+
     iget-object p0, p0, Lcom/android/server/wm/ActivityTaskManagerService;->mTaskChangeNotificationController:Lcom/android/server/wm/TaskChangeNotificationController;
 
-- 
2.55.0
